compiler-core/src/analyse/name.rs

```diff
@@ -0,0 +1,139 @@
+use ecow::{eco_format, EcoString};
+use regex::Regex;
+
+use crate::ast::{ArgNames, SrcSpan};
+
+use super::{BadNameKind, Error};
+use heck::{ToSnakeCase, ToUpperCamelCase};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NameCorrection {
+    pub location: SrcSpan,
+    pub correction: EcoString,
+}
+
+pub fn check_valid_name(
```

Comment:
These functions do more than just check the name, they also return a correction. Let's move that out of these and keep them doing one thing.

Let's also make the error case `Err`.